File: clients/java/src/main/java/mo/harness/Result.java
package mo.harness;

/** Outcome of running a scenario. */
public final class Result {
    public final String framework;
    public final String category;
    public final String name;
    public final String status;       // PASS | FAIL | SKIP
    public final long durationMs;
    public final String errorCode;    // numeric MO code, "BEHAVIOR", SQLState, or null
    public final String errorMessage; // truncated

    public Result(String framework, String category, String name, String status,
                  long durationMs, String errorCode, String errorMessage) {
        this.framework = framework;
        this.category = category;
        this.name = name;
        this.status = status;
        this.durationMs = durationMs;
        this.errorCode = errorCode;
        this.errorMessage = errorMessage;
    }
}

File: clients/python/harness/result.py
"""Scenario result types and control-flow exceptions."""

from __future__ import annotations

from dataclasses import dataclass


class SkipScenario(Exception):
    """Raise to mark a scenario SKIP (e.g. an unrelated prerequisite is absent)."""


class BehaviorMismatch(Exception):
    """Raise when a statement ran without error but the result differs from MySQL.

    These are recorded as FAIL with error_code 'BEHAVIOR' — the most dangerous
    class of incompatibility because there is no error to alert the application.
    """


PASS = "PASS"
FAIL = "FAIL"
SKIP = "SKIP"


@dataclass
class TestResult:
    id: int
    framework: str
    category: str
    name: str
    status: str
    duration_ms: float
    detail: str | None = None
    error_code: str | None = None
    error_message: str | None = None

    def to_dict(self) -> dict:
        msg = self.error_message
        if msg is not None and len(msg) > 200:
            msg = msg[:197] + "..."
        return {
            "framework": self.framework,
            "category": self.category,
            "name": self.name,
            "status": self.status,
            "error_code": self.error_code,
            "error_message": msg,
        }
